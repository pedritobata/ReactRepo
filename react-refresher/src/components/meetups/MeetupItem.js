import React, {useContext} from 'react';
import FavoritesContext from '../../store/favorites-context';
import classes from './MeetupItem.module.css';


const MeetupItem = (props) => {
    const {addFavorite, removeFavorite, isFavorite} = useContext(FavoritesContext);

    const toggleFavoritesHandler = () => {
        if(isFavorite(props.id)){
            removeFavorite(props.id);
        } else {
            addFavorite({
                id: props.id,
                title: props.title,
                image: props.image,
                address: props.address,
                description: props.description
            })
        }
    }

    return (
        <li className={classes.item}>
            <div className={classes.image}>
                <img src={props.image} alt={props.title} />
            </div>
            <div className={classes.content}>
                <h3>{props.title}</h3>
                <address>{props.address}</address>
                <p>{props.description}</p>
            </div>
            <div className={classes.actions}> 
                <button onClick={toggleFavoritesHandler}>{isFavorite(props.id) ? 'Remove from favorites' : 'To Favorites ðŸ’–'}</button>
            </div>
        </li>
    )
}

export default MeetupItem;