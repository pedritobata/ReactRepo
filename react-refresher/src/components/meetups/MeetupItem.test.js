import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import MeetupItem from './MeetupItem';
import FavoritesContext from '../../store/favorites-context';



describe('MeetupItem', () => {
    it('should render correctly', () => {
        const snap = renderer.create(<MeetupItem />).toJSON();
        expect(snap).toMatchSnapshot();
    })

    it('should call proper method when removing favorites', () => {
        const mockAddFavorite = jest.fn();
        const mockRemoveFavorite = jest.fn();
        const mockIsFavorite = jest.fn();
        mockIsFavorite.mockReturnValueOnce(true).mockReturnValueOnce(true).mockReturnValueOnce(false);
        const mockContext = {
            addFavorite: mockAddFavorite, 
            removeFavorite: mockRemoveFavorite,
            isFavorite: mockIsFavorite
        }
        const WrapperCtx = ({children}) => <FavoritesContext.Provider value={mockContext}>{children}</FavoritesContext.Provider>;
        const props = {
            id: 'fav1',
            title: 'first meetup',
            image: 'http://somepath/image.jpeg',
            address: 'Trinidad Moran',
            description: 'Que nadie venga a joder!!'
        }

        const {rerender} = render(<WrapperCtx><MeetupItem  {...props}/></WrapperCtx>);

        expect(mockIsFavorite).toHaveBeenCalledWith(props.id);
        expect(mockIsFavorite.mock.results[0].value).toBe(true);
        const buttonFav = screen.getByRole('button');
        expect(buttonFav).toHaveTextContent('Remove from favorites');
        userEvent.click(buttonFav);
        expect(mockRemoveFavorite).toHaveBeenCalled();
        rerender(<WrapperCtx><MeetupItem  {...props}/></WrapperCtx>);
        expect(mockIsFavorite.mock.results[2].value).toBe(false);
        expect(buttonFav).toHaveTextContent('To Favorites');

    })
})