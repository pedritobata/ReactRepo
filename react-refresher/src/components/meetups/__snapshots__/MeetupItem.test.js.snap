// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeetupItem should render correctly 1`] = `
<li
  className="item"
>
  <div
    className="image"
  >
    <img />
  </div>
  <div
    className="content"
  >
    <h3 />
    <address />
    <p />
  </div>
  <div
    className="actions"
  >
    <button
      onClick={[Function]}
    >
      To Favorites ðŸ’–
    </button>
  </div>
</li>
`;
